### DISCUSSION TITLE
Use bootstrap rows and columns in "Shopping Experiences"

### DISCUSSION THREAD

**Discussion by aragon999** [shopware/platform#3184] [RFC / Proposal]
Effort: high

# Description 
## Problem

Currently there exists the block system for the "Shopping Experiences"-layouts, which is quite limited, in terms adding and removing different columns, with different sizes.

For example if one wants to have three columns, on the left there are two text columns and on the right there is one image column all equally sized, one can add the "Three columns, text" block and replace the leftmost text element by an image element:
![image](https://github.com/shopware/platform/assets/6317761/3fb936da-306b-473b-ba44-607b0840151e)
But if I want to have an element which has only two columns, one text column on the right and one image column on the left, but the text should span over two columns, it is not possible without adding a custom block, which has such a layout. On the other hand if I would have an image on the left (width 1/3) and a text on the right (width 2/3), one can realize such a combination using "Two columns, teaser text" and replace the left element by an image element:
![image](https://github.com/shopware/platform/assets/6317761/ad258998-7e42-48d2-ac95-d855f4b07170)

Furthermore as a plugin developer I need to ensure that my elements can be replaced, otherwise these cannot be used to replace existing elements.

## Solution

The block system should be removed and replaced by a system which allows to define (Bootstrap) columns. 

In more detail, for each "section", one can define, a gapped or a gapless, Bootstrap row. For this row one can than define how many columns should be used, each with their own classes. There should probably exist some predefined layouts, maybe 50:50 columns, 1/4:3/4 columns, 1/3:1/3:1/3 columns, and so on. Additionally one can use the "expert mode" to freely assign the Bootstrap column classes for each viewport separately.

The elements should than be listed without blocks, which should also reduce the number of elements a lot. For example for the text and image elements, only the heading element, text element and a single image element is needed. Additionally a image gallery would also be needed as an element. These four elements are a lot less in comparison to the 17 elements, which can be replaced by those in combination with the system described.

To each column one can add an arbitrary amount of elements, which are than listed one after another. Therefore each element should have the possibility to add a margin or padding for the different viewports.

Maybe it should also be possible to adjust the number of columns of an existing row, where when reducing the number, the elements of the removed columns are all moved into the last remaining column.

# Benefits

- Fewer Shopping experience elements
- More customizations to style Shopping experience layouts, without having to change code

# Break strategy

- Each existing block should be automatically converted to the correspondent row/column layout
- It should not be possible to add blocks but only elements
- Maybe a hybrid more is needed, where both functionalities can be used side by side

**Comment by jankal** [comment]
From my understanding, it would be "cheaper" effort wise, for Shopware to ship with all kinds of possible column grid layout blocks.

Those are:
- col-12
- col-11 : col-1
- col-10 : col-2
- col-10 : col-1 : col-1
- col-9 : col-3
- col-9 : col-1 : col-2
- col-9 : col-2 : col-1
- col-9 : col-1 : col-1 : col-1
- col-8 : col-4
- col-8 : col-2 : col-2
- col-8 : col-1 : col-1 : col-2
- col-8 : col-3 : col-1
- col-8 : col-1 : col-3
- col-8 : col-2 : col-1 : col-1
- col-8 : col-1 : col-1 : col-1 : col-1
- ...

In the end, this approach comes down to some amount of manual labor once in SW core, but wouldn't have implications to the whole ecosystem.

Another method would be to build a "dynamic" configurable grid block (not labor-intensive but probably quite high complexity)

**Reply by runelaenen** [reply]
While easier to implement, having tens/hundreds of blocks with different 'ready' columns configurations does not seem to be a good solution.

There are plugins available (for example my own grid plugin https://store.shopware.com/en/laene34427173466/cms-bootstrap-grid-section.html ) that implement the 'hybrid' approach to this issue. So it is certainly doable to implement a configurable way of working.
I would be a supporter of implementing a similar hybrid way of working in the core. This builds further on the existing system, making sure it stays compatible with existing block-pages.

**Reply by jankal** [reply]
I didn't know about your plug-in @runelaenen. Actually I like the idea of a hybrid approach instead of exchanging current containers with a grid layout and having asteroids migration path.

What do you think @aragon999?

**Reply by aragon999** [reply]
I also did not know about the plugin, sorry @runelaenen it was not meant to make the plugin obsolete :-)

But I also like this approach, tbh also more than having predefined column layouts. Just at some point the old blocks should probably be deprecated and removed. 

One still has to think about how elements from plugins are handled, since as a plugin manufacturer, we need to somehow ensure that all our elements can also be used standalone and not only within a block of the plugin.

But unfortunately I am not really that much of a JS/frontend developer, so I might miss something here. I just find it quite hard and not flexible to create layouts in the current form.

**Reply by runelaenen** [reply]
I did not mean my answer as a bad thing. I just offered it as an example of how it could be done 😄

> But I also like this approach, tbh also more than having predefined column layouts. Just at some point the old blocks should probably be deprecated and removed.

I agree

> One still has to think about how elements from plugins are handled, since as a plugin manufacturer, we need to somehow ensure that all our elements can also be used standalone and not only within a block of the plugin.

In theory, if implemented properly, an element should not depend on which block it is placed in. However there are indeed CMS elements that will probably only work in certain blocks currently.

Also some elements will not look as good if put into a 'col-1'. However, that is more a matter of 'educating' the user I think to be using sensible column configurations.

**Comment by Phil23** [comment]
Hey @aragon999, thanks for the feedback. We have exactly that in mind. 👍 

We already mentioned it briefly here:
https://github.com/shopware/shopware/discussions/4856

But we can leave this discussion open to discuss this specific feature in detail to find the best solution. I see that there are already some comments with ideas about how to implement it. Our current idea is to create a configurable block element that can make use of all the features of the Bootstrap grid system. This will be part of the new component system that we are planning.
